Mathlib/Tactic/LieAlgebra/LieRingNF.lean

```diff
@@ -0,0 +1,222 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nailin Guan, Zixun Guo, Wanyi He, Jingting Wang
+-/
+import Mathlib.Tactic.LieAlgebra.Basic
+import Mathlib.Tactic.Module
+
+/-!
+# `lie_ring_nf` tactic
+
+A tactic that uses `lie_ring` to simplify expressions to Lyndon normal form.
+It's expected that this tactic is one of the building blocks to the `lie_algebra` tactic.
+
+The implementation of this tactic follows the `ring_nf` tactic.
+-/
+
+open Lean Meta Elab Mathlib.Tactic Qq
+
+namespace Mathlib.Tactic
+
+open Mathlib.Tactic.LieRing
+
+namespace LieRingNF
+
+private inductive LieRingConfig
+  | simple
+  | raw
+deriving Inhabited, BEq, Repr
+
+private structure Config where
+  /-- the reducibility setting to use when comparing atoms for defeq -/
+  red := TransparencyMode.reducible
+  /-- if true, atoms inside ring expressions will be reduced recursively -/
+  recursive := true
+  /-- the strategy to use for normalizing the lie ring -/
+  strategy : LieRingConfig := LieRingConfig.simple
+  deriving Inhabited, BEq, Repr
+
+/--
+  Default elaborator
+-/
+declare_config_elab elabConfig Config
+
+private structure Context where
+  ctx : Simp.Context
+  simp : Simp.Result → SimpM Simp.Result
+
+private abbrev M := ReaderT Context AtomM
+
+private partial def rewrite (parent : Expr) (root := true) : M Simp.Result :=
+  fun nctx rctx s ↦ do
+    let pre : Simp.Simproc := fun e =>
+      try
+        guard <| root || parent != e -- recursion guard
+        let e ← withReducible <| whnf e
+        guard e.isApp -- all interesting expressions are applications
+        let ⟨u, α, e⟩ ← inferTypeQ' e
+        let sα ← synthInstanceQ (q(LieRing $α) : Q(Type u))
+        let ⟨a, _, pa⟩ ← if ← isAtom α e rctx s then
+          (failure : MetaM (Result (ExSum sα) e)) -- No point rewriting atoms
+        else
+          -- notice that in our design,
+          -- when we come across `u • ⁅a, b⁆ + v • ⁅c, d⁆`, we pass `⁅a, b⁆` and `⁅c, d⁆` to `eval`
+          -- separately, rather than the whole expr.
+          -- eval sα e rctx s -- `none` indicates that `eval` will find something algebraic.
+          let .const n _ := (← withReducible <| whnf e).getAppFn | failure
+          match n with
+          | ``Bracket.bracket =>
+            eval sα e rctx s
+          | _ =>
+            -- if it is not a lie bracket, recursively rewrite the arguments
+            -- after failure, the simp process automatically continues into subexpressions
+            failure
+        let r ← nctx.simp { expr := a, proof? := pa }
+        if ← withReducible <| isDefEq r.expr e then return .done { expr := r.expr }
+        pure (.done r)
+      catch _ => pure <| .continue
+    let post := Simp.postDefault #[]
+    (·.1) <$> Simp.main parent nctx.ctx (methods := { pre, post })
+
+section SimpLemmas
+variable {R L : Type*} [CommRing R] [LieRing L] [LieAlgebra R L] {n d : ℕ}
+
+private theorem zero_smul1 (a : L) : (0 : R) • a = 0 := zero_smul _ _
```

Comment:
Why do you explicitly need to add all these lemmas that can at any rate be found verbatim? I understand that you don't want complicated ones in the middle of your meta-code, but what's the difference between calling `zero_smul` or `zero_smul1`?

Reply:
I think the idea is to produce shorter proof terms, and to make sure that only `smul`s relating to lie rings are simplified.